### PR TITLE
[eQSL] Default QSLMSG by station profil

### PR DESCRIPTION
**Pull request by abarrau**

### CODE REVIEW

**Comment by abarrau** [comment]
Fonctionnality Default QSLMSG : 
- on station profil (add and create), user can defined a text for this station. 
- on new QSO (live, post) : 
  - text is automatically completed on qslmsg esql field
  - if user change station, the default of this new station rewritten the new text
  - but if user has change the default text during the QSO, the new text is conserved
  - a reload icon (near title) allow to reload the default text of current station
- on new QSO (Contest or SFL)
  - the default qslmsg is define at the time of save qso
- on edit QSO, same fonctionnality
  - rewritte new text or reload default text
- default qslmsg et qslmsg are limited to 240 chars

Station profil page : 
station profil page

QSO (with icon) : 
qso page

File: application/controllers/Qso.php [hunk 1]
```diff
@@ -119,6 +120,14 @@ public function index()
 	 * This is used for contest-logging and the ajax-call
 	 */
 	public function saveqso() {
+		// search qslmsg by default for this station (when contest or SFL) overload qslmsg post value if empty //
+		if (trim($this->input->post('qslmsg'))=='') {
+			$this->load->model('stations');
+			if (!(intval($this->input->post('station_profile'))>0)) {
+				$_POST['station_profile'] = $this->stations->find_active();
+			}
+			$_POST['qslmsg'] = $this->stations->get_station_info($_POST['station_profile'],'eqsl_defaultqslmsg', '');
```
Comment (int2001) [inline application/controllers/Qso.php:129]: Why assigning values to $_POST?
$_POST is a var which contains the posted things. it's there to READ, not to transport data within the application.

File: application/controllers/Qso.php [hunk 2]
```diff
@@ -29,7 +29,8 @@ public function index()
         if(!$this->user_model->authorize(2)) { $this->session->set_flashdata('notice', 'You\'re not allowed to do that!'); redirect('dashboard'); }
 
 		$data['active_station_profile'] = $this->stations->find_active();
-        
+		$data['qslmsg'] = $this->stations->get_station_info($data['active_station_profile'],'eqsl_defaultqslmsg', '');
```
Comment (int2001) [inline application/controllers/Qso.php:32]: Why using (new) get_station_info where a function called "profile" alrteady exists?

File: application/controllers/Qso.php (hunk 1 above)
Comment (int2001) [reply application/controllers/Qso.php:129]: Further: Please check if station_id belongs to user! Otherwise there are security-risks. (Yes, i saw that you implemented it in the redundant function "get_station_info" at Stations-Model - but no need to implement redundant things here)

File: application/controllers/Station.php [hunk 3]
```diff
@@ -208,4 +208,17 @@ public function get_county() {
...
 
+    // Function return db value from station_profil field //
+	public function get_json_info() {
+		$return_json = array();
+		$this->load->model('stations');
+		$station_id = $this->input->post('station_id'); 
+		$field = $this->input->post('field');
+		if (($station_id>0) && (!empty($field))) {
+			$return_json[$field] = $this->stations->get_station_info($station_id,$field,"");
+		}
```
Comment (int2001) [inline application/controllers/Station.php:219]: See comment at Qso.php.
- get_station_info is not nescessary. it's redundant to profile.
- Please CHECK that station_id belongs to user
- dynamic-fields ($field) is not a good idea. If User/Attacker puts in "qrzapikey" as field, it'll be exposed. Security-Risk!

File: application/migrations/153_add_eqsldefaultmsg.php
Comment (int2001) [inline application/migrations/153_add_eqsldefaultmsg.php:1]: No need for migration.
Please check user_options. There's enough space to put a custom eQSL-Message.

File: application/models/Stations.php [hunk 4]
```diff
@@ -496,6 +498,19 @@ public function check_station_against_callsign($stationid, $callsign) {
 		}
 		return false;
 	}
+
+	// Function return db value from station_profil field //
+	public function get_station_info($id,$field=null,$default_return=false) {
```
Comment (int2001) [inline application/models/Stations.php:503]: whole function is redundant to "profile". I see you implemented a "whitelist" so only eqsl_defaultqslmsg is allowed. But please use "profile"

**Review by int2001 — CHANGES_REQUESTED** [review]
See comments.
I like the idea behind that feature, but please check code. tnx!

File: application/controllers/Qso.php (hunk 1 above)
Comment (abarrau) [reply application/controllers/Qso.php:129]: Hello, 
1/ i use $_POST[] because : 
- $_POST['station_profile'] : not exist when a contest call saveqso() ; the station_profile is only defined in function create_qso() on logbook_model ; first time i would to add my test in this function but if user must an empty value qslmsg, i can not defined if a real empty value or it the qso create by SFL (SFL don't send  station_profile too ; even if we add a new information in ajax call in function js-save-to-lol(), like "isSFLQso" ?)
- $_POST['qslmsg'] : because it's not send by contest and SFL QSO save

and fonction logbook_model->create_qso(), use only information form POST, how can i send information to create_qso(), if i dont overload POST :(

2/ ok i can use check_station_is_accessible()

File: application/controllers/Qso.php (hunk 2 above)
Comment (abarrau) [reply application/controllers/Qso.php:32]: profile return all data, i think to create a function like get_station_power(), but more generic. 
ok for use profile() if you prefered

File: application/controllers/Station.php (hunk 3 above)
Comment (abarrau) [reply application/controllers/Station.php:219]: field is not dynamic, it was checked on exist field in table  `$fields_allow = array('eqsl_defaultqslmsg');`
with this method, we can avoid using several functions for the same final action (return 1 value).
But i wil change to profile(), so

File: application/migrations/153_add_eqsldefaultmsg.php
Comment (abarrau) [reply application/migrations/153_add_eqsldefaultmsg.php:1]: with user_options, how can i add qslmsg by station ? 
i can propose : 
- option_name = eqsl_defautlqslmsg_<userid>_<stationid>
- option_value = txt ... 
this way is ok for you ?

Comment (int2001) [reply application/migrations/153_add_eqsldefaultmsg.php:1]: yes. you have 3 "keys" u can use.
i suggest: 
- "default_eqslmessage" (or sth. like that) for option_type
- option_name could be "station_eqsl_message"
- option_key the station_id
- option_value the message itself.

File: application/controllers/Station.php (hunk 3 above)
Comment (int2001) [reply application/controllers/Station.php:219]: yes, that was the idea behind "profile". it fetches all at the model-layer. and at controller you can specify which one is needed and if the key is granted and so on.
the power-thing is a historic relict ;)

File: application/controllers/Qso.php (hunk 1 above)
Comment (int2001) [reply application/controllers/Qso.php:129]: you're right.
what about another way: prepopulate the eQSL-Msg-Field at the View upon loading the Form? I see more advantages here:
- User can change if he wants to type sth. special in it
- no need to set $_POST

File: application/migrations/153_add_eqsldefaultmsg.php
Comment (abarrau) [reply application/migrations/153_add_eqsldefaultmsg.php:1]: I didn't have the same structure on my bdd : 

option table

option_type and option_key not exist ...

File: application/controllers/Qso.php (hunk 1 above)
Comment (abarrau) [reply application/controllers/Qso.php:129]: This is what is done in the context of QSOs (live/post), but in the case of a contest or SFL, the field does not exist...
I can add it by hiding it for contests, but not for SFL...
ok I have an idea, I will submit it

File: application/migrations/153_add_eqsldefaultmsg.php
Comment (abarrau) [reply application/migrations/153_add_eqsldefaultmsg.php:1]: i dont know you prefered options table than station table
the code will be harder to implement it 
tell me what the team prefers

Comment (int2001) [reply application/migrations/153_add_eqsldefaultmsg.php:1]: before we implemented user_options every table (like station_profiles, users, etc.) has grown dramatically because of flags and settings. that's the reason for user_options. it's still on the internal roadmap to migrate some flags to that table. But new ones should use it.

if option_key ist missing at your DB, there's probably a problem with migration-scripts on your system. migration-script 138 created the table with the option_key field

Comment (abarrau) [reply application/migrations/153_add_eqsldefaultmsg.php:1]: ok, it's user_options , not options, sorry ... 
user option table

So i will recoding with this table ;) ...

Comment (int2001) [reply application/migrations/153_add_eqsldefaultmsg.php:1]: check out the model for user_options. it contains everything you need

**Comment by abarrau** [comment]
@int2001 , 
I change my method to create/ask default qslmsg. 
hope is ok for you now.

File: application/models/User_options_model.php [hunk 5]
```diff
@@ -16,16 +16,32 @@ public function set_option($option_type, $option_name, $option_array) {
...
-	public function del_option($option_type, $option_name) {
+	public function del_option($option_type, $option_name, $option_array=null) {
 		$uid=$this->session->userdata('user_id');
-		$sql='delete from user_options where user_id=? and option_type=? and option_name=?';
-		return $this->db->query($sql, array($uid, $option_type,$option_name));
+		$sql_more = "";
+		$array_sql_value = array($uid, $option_type, $option_name);
+		if (is_array($option_array)) {
+			foreach ($option_array as $key => $value) {
+				$sql_more .= ' and '.$key.'=?';
```
Comment (int2001) [inline application/models/User_options_model.php:39]: please continue the use of the binding array.
Rest here looks good

**Comment by int2001** [comment]
much better than the first approach. congrats.

a few things:
- can you pull the latest dev? there's an error at live-qso which was fixed by @phl0 - it isn't included here (and will be reverted, if this one gets merged)
- please note my comment about binding SQL-Vars (much safer than inline literals)

Tested so far:
- Setting default-Comment. Works
- Logging live. Works (except the above mentioned error)
- Logging Post. Works (except the above mentioned error)
- Logging contest

open for testing:
- SFLE (@HB9HIL ?)
- ... ?

from security-aspects this PR looks fine to me. to be honest: flew over the code and noticed nothing evil, yet.

File: application/models/User_options_model.php (hunk 5 above)
Comment (abarrau) [reply application/models/User_options_model.php:39]: Sorry, i don't uderstand where is the problem here ?

**Comment by abarrau** [comment]
hello @int2001 , thanks for your return. 

Where i can find this fix ? 
`can you pull the latest dev? there's an error at live-qso which was fixed by @phl0 - it isn't included here (and will be reverted, if this one gets merged)`
I synchronized my branch with dev, but i didn't show difference :(

**Comment by int2001** [comment]
you merged master, not dev...

**Comment by abarrau** [comment]
> you merged master, not dev...

Sorry, I admit my ignorance, I can't find how to resynchronize with dev, it systematically offers me master :(
(and since I only do stupid things I'm still afraid of doing one...)

**Comment by int2001** [comment]
try the following:
```
git checkout dev
git pull
git checkout ans-eqsl-default-qslmsg
git merge dev
```
at that point there will occure some merge-problems. you've to resolve them manually.
after that is done, you should test everything and commit with
`git commit`
then push it to github with
`git push`

**Comment by abarrau** [comment]
Hello, 
Thanks for your help (but i not use commande command line), i hope it's ok now. 
You make a new check if ok for you !
73

**Comment by HB9HIL** [comment]
> 
> open for testing:
> 
> * SFLE (@HB9HIL ?)
> * ... ?
> 

I didn't  check the code yet. But SFLE is binded to a Station Location. If the QSL MSG to eQSL is trigger by a script which is only active in qso view, then it wont work. If its binded to the Station Location and gets triggerd by the eqsl cron job it will probably work.

File: application/models/User_options_model.php (hunk 5 above)
Comment (int2001) [reply application/models/User_options_model.php:39]: looks good

**Review by int2001 — APPROVED** [review]
@HB9HIL pse check/review, too.
SFLE - you know.

From my side it looks ok.

**Review by HB9HIL — CHANGES_REQUESTED** [review]
Okay. I reviewed the code aswell, especially the SFLE Part. Looks nearly good to me. Just maybe one thing for esthetic reasons. 

Rename the Var `isSFL` to `isSFLE`

in

`js/sections/simplefle.js Line 763`
`application/models/Logbook_model.php Line 264`

Reason is to have the same name for that in the Code (we use "simplefle" and "sfle").

Good Job! :)

**Comment by abarrau** [comment]
Sorry, i think i make a mistake with my last merge ... :(

**Comment by abarrau** [comment]
i close it and make a new PR next week ... (with @HB9HIL assistance :))
sorry, team !
